File: 208-implement-trie/solution.cpp
#include <string>
#include <unordered_map>
#include <unordered_set>

struct TrieNode
{
public:
  TrieNode(const char c = '\0', const TrieNode *const parent = nullptr) : parent{parent}, val{parent == nullptr ? "" : (parent->val + c)} {}

  std::unordered_map<char, TrieNode *> edges;
  const TrieNode *const parent = nullptr;
  const std::string val;
};

class Trie
{
private:
  TrieNode *root = new TrieNode();
  std::unordered_set<std::string> words;

public:
  Trie() {}

  void insert(std::string word)
  {
    this->words.emplace(word);
    auto node = this->root;
    for (const char c : word)
    {
      auto it = node->edges.find(c);
      if (it == node->edges.end())
      {
        auto result = node->edges.emplace(c, new TrieNode(c, node));
        it = result.first;
      }

      node = it->second;
    }
  }

  bool search(std::string word)
  {
    return this->words.contains(word);
  }

  bool startsWith(std::string prefix)
  {
    const auto *node = this->root;
    for (const char c : prefix)
    {
      auto it = node->edges.find(c);
      if (it == node->edges.end())
      {
        return false;
      }

      node = it->second;
    }

    return true;
  }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: 208-implement-trie/solution.py
from typing import Optional


class TrieNode:
  def __init__(self, char='', parent: Optional['TrieNode'] = None):
    self.edges: dict[str, 'TrieNode'] = {}
    self.parent = parent
    self.val = parent.val if parent is not None else '' + char


class Trie:

  def __init__(self):
    self.root = TrieNode()
    self.words: set[str] = set()

  def insert(self, word: str) -> None:
    self.words.add(word)
    node = self.root
    for char in word:
      if char not in node.edges:
        node.edges[char] = TrieNode(char, node)

      node = node.edges[char]

  def search(self, word: str) -> bool:
    return word in self.words

  def startsWith(self, prefix: str) -> bool:
    node = self.root
    for char in prefix:
      if char not in node.edges:
        return False

      node = node.edges[char]

    return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
